### PR TITLE
[WIP] Tab presenter for WPF and update Playground to demonstrate

### PR DESCRIPTION
**Pull request by Happin3ss**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)

Add tab presenter for WPF and update Playground to demonstrate
 Original things from @benhysell . I made some changes for it to work with v6 and also can save the view/viewmodel states

### :arrow_heading_down: What is the current behavior?
No tab presenter for WPF

### :new: What is the new behavior (if this is a feature change)?
New tab presenter for WPF

### :boom: Does this PR introduce a breaking change?
Yes, but only for Playground

### :memo: Links to relevant issues/docs
https://github.com/MvvmCross/MvvmCross/issues/2712

### :thinking: Checklist before submitting

- [x] All projects build
- [ ] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [ ] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contribute/mvvmcross-docs-style-guide))
- [x] Rebased onto current develop

### CODE REVIEW

File: MvvmCross/Platforms/Wpf/Presenters/Attributes/MvxRegionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+
+namespace MvvmCross.Platforms.Wpf.Presenters.Attributes
+{
+	[AttributeUsage(AttributeTargets.Class, Inherited = false)]
+	public class MvxRegionAttribute
```
Comment (martijn00) [inline MvvmCross/Platforms/Wpf/Presenters/Attributes/MvxRegionAttribute.cs]: Name should be `MvxRegionPresentationAttribute`

File: MvvmCross/Platforms/Wpf/Presenters/Attributes/MvxRegionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+		: Attribute
```
Comment (martijn00) [inline MvvmCross/Platforms/Wpf/Presenters/Attributes/MvxRegionAttribute.cs]: The attribute needs to extend `MvxBasePresentationAttribute`

**Comment by Happin3ss** [comment]
@benhysell Please take a look on this one. It's all based on your works before.
@All : I made changes to playground just only for demonstrating purpose. It need to be refactor to fit with the current playground.

File: MvvmCross/Platforms/Wpf/Presenters/MultiRegionWpfViewPresenter.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Windows;
+using System.Windows.Controls;
+using MvvmCross.Platforms.Wpf.Views;
+using MvvmCross.ViewModels;
+using MvvmCross.Views;
+
+namespace MvvmCross.Platforms.Wpf.Presenters
+{
+    public class MultiRegionWpfViewPresenter : MvxWpfViewPresenter
```
Comment (martijn00) [inline MvvmCross/Platforms/Wpf/Presenters/MultiRegionWpfViewPresenter.cs]: We don't want multiple presenters anymore. Can you integrate this into the normal presenter?

File: MvvmCross/Platforms/Wpf/Presenters/MvxWpfExtensionMethods.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Linq;
+using MvvmCross.Platforms.Wpf.Presenters.Attributes;
+
+namespace MvvmCross.Platforms.Wpf.Presenters
+{
+    public static class MvxWpfExtensionMethods
```
Comment (martijn00) [inline MvvmCross/Platforms/Wpf/Presenters/MvxWpfExtensionMethods.cs]: It seems this only contains Presenter extensions. Please name it `MvxWpfPresenterExtensions`

File: MvvmCross/Platforms/Wpf/Presenters/MvxWpfPage.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿using System.Windows.Controls;
+using MvvmCross.Platforms.Wpf.Views;
+using MvvmCross.ViewModels;
+
+namespace MvvmCross.Platforms.Wpf.Presenters
```
Comment (martijn00) [inline MvvmCross/Platforms/Wpf/Presenters/MvxWpfPage.cs]: This should be in the `Views` folder. Also it might need some more work, see: https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross/Platforms/Wpf/Views/MvxWpfView.cs

File: Projects/Playground/Playground.Core/ViewModels/RootViewModel.cs [hunk 6]
```diff
@@ -59,6 +59,12 @@ public override void Prepare()
             base.Prepare();
         }
 
+        public void ShowMenu()
```
Comment (martijn00) [inline Projects/Playground/Playground.Core/ViewModels/RootViewModel.cs]: Should be `Task` and `await` navigate calls.

File: Projects/Playground/Playground.Wpf/App.xaml.cs [hunk 7]
```diff
@@ -1,10 +1,24 @@
 ﻿using MvvmCross.Core;
 using MvvmCross.Platforms.Wpf.Core;
+using MvvmCross.Platforms.Wpf.Presenters;
 using MvvmCross.Platforms.Wpf.Views;
 
 namespace Playground.Wpf
 {
-    public partial class App : MvxApplication<MvxWpfSetup<Core.App>, Core.App>
+    public partial class App : MvxApplication
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/App.xaml.cs]: Why is this changed?

File: Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs [hunk 8]
```diff
@@ -1,15 +1,27 @@
...
 using Playground.Core.ViewModels;
 
 namespace Playground.Wpf.Views
 {
 
     [MvxViewFor(typeof(RootViewModel))]
-    public partial class RootView 
+    public partial class RootView
     {
+        public new RootViewModel ViewModel
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs]: Us a generic instead of this way

File: Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs [hunk 9]
```diff
@@ -1,15 +1,27 @@
...
+        {
+            get { return (RootViewModel)base.ViewModel; }
+            set { base.ViewModel = value; }
+        }
         public RootView()
         {
             InitializeComponent();
+            Loaded += MainView_Loaded;
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs]: Can't this be done from the `ViewModel.ViewAppeared`?

File: Projects/Playground/Playground.Wpf/Views/Tab1View.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+using MvvmCross.Platforms.Wpf.Presenters.Attributes;
+using Playground.Core.ViewModels;
+
+namespace Playground.Wpf.Views
+{
+    [MvxRegion("Tab1Content")]
+
+    public partial class Tab1View : MvxWpfPage
+    {
+        public new Tab1ViewModel ViewModel
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/Views/Tab1View.xaml.cs]: Use a generic

File: Projects/Playground/Playground.Wpf/Views/Tab2View.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+using Playground.Core.ViewModels;
+
+namespace Playground.Wpf.Views
+{
+    [MvxRegion("Tab2Content")]
+    public partial class Tab2View : MvxWpfPage
+    {
+        public new Tab2ViewModel ViewModel
+        {
+            get { return (Tab2ViewModel)base.ViewModel; }
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/Views/Tab2View.xaml.cs]: Use a generic

File: Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs [hunk 12]
```diff
@@ -1,15 +1,27 @@
-﻿using MvvmCross.ViewModels;
+﻿using System.Windows;
+using MvvmCross.ViewModels;
 using Playground.Core.ViewModels;
 
 namespace Playground.Wpf.Views
 {
 
     [MvxViewFor(typeof(RootViewModel))]
```
Comment (martijn00) [inline Projects/Playground/Playground.Wpf/Views/RootView.xaml.cs]: I think we might need to refactor `MvxViewFor` into the presentation attributes

**Comment by martijn00** [comment]
@Happin3ss are you able to work on this?

**Comment by Happin3ss** [comment]
@martijn00 hello Martin, I'm kind of busy in the previous month. I will continue on tmr !

**Comment by Happin3ss** [comment]
Currently I'm still on a very busy phase with a lot of travelling so I cannot continue this one as I expected. Please consider this is on hold and feel free to contribute/finish it. Thank you so much !

**Comment by nickrandolph** [comment]
@Happin3ss are you still busy with other work? Any chance you can jump back onto this PR so we can progress it?

**Comment by nickrandolph** [comment]
@Happin3ss we're really like to progress this PR - are you available to rebase and work out what still needs to be done to push this PR through?

**Comment by yanxiaodi** [comment]
Hi, @martijn00 @nickrandolph , if @Happin3ss is too busy to do it, can I continue to work on it? I think the tab presenter for WPF is a good idea. Thanks.

**Comment by Cheesebaron** [comment]
Sure go ahead. The code in this PR might need a rebase.

**Comment by yanxiaodi** [comment]
@Cheesebaron should I continue to work on my fork then create a new PR? I'm not sure if I can change the current branch. Thanks.

**Comment by Cheesebaron** [comment]
You will probably need to create a new PR and your own branch for that.

**Comment by Cheesebaron** [comment]
Closing this one as stale.
